File: chapters/10_security_sandbox_pairing/v2_policy_pairing.py
from __future__ import annotations


def enforce_dm_policy(
    is_dm: bool,
    user: str,
    dm_policy: str,
    paired_users: set[str],
) -> tuple[str, str]:
    if not is_dm:
        return "allowed", "non-dm"

    if dm_policy == "allow_all":
        return "allowed", "dm-allow-all"

    if dm_policy == "paired_only" and user in paired_users:
        return "allowed", "dm-paired"

    return "blocked", "dm-policy"


def main() -> None:
    decision, reason = enforce_dm_policy(True, "alice", "paired_only", {"bob"})
    print(decision, reason)


if __name__ == "__main__":
    main()

File: chapters/10_security_sandbox_pairing/v2_policy_pairing.ts
// TypeScript mirror for Chapter 10: Security Sandbox Pairing (v2)
// Run with: npx ts-node v2_policy_pairing.ts

function enforceDmPolicy(
  isDm: boolean,
  user: string,
  dmPolicy: string,
  pairedUsers: Set<string>
): [string, string] {
  if (!isDm) {
    return ["allowed", "non-dm"];
  }

  if (dmPolicy === "allow_all") {
    return ["allowed", "dm-allow-all"];
  }

  if (dmPolicy === "paired_only" && pairedUsers.has(user)) {
    return ["allowed", "dm-paired"];
  }

  return ["blocked", "dm-policy"];
}

function main(): void {
  const [decision, reason] = enforceDmPolicy(true, "alice", "paired_only", new Set(["bob"]));
  console.log(decision, reason);
}

main();
